sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -128,26 +133,34 @@ public CompletableResultCode forceFlush() {
     private final long scheduleDelayNanos;
     private final int maxExportBatchSize;
     private final long exporterTimeoutNanos;
-
     private long nextExportTime;
-
-    private final BlockingQueue<ReadableSpan> queue;
-
+    private final ConcurrentLinkedQueue<ReadableSpan> queue;
+    private final int maxQueueSize;
+    private final AtomicInteger queueSize = new AtomicInteger(0);
+    private final AtomicLong addedSpansCounter = new AtomicLong(0);
     private final AtomicReference<CompletableResultCode> flushRequested = new AtomicReference<>();
     private volatile boolean continueWork = true;
     private final ArrayList<SpanData> batch;
+    private final ReentrantLock lock;
+
+    @GuardedBy("lock")
+    private final Condition needExport;
 
     private Worker(
         SpanExporter spanExporter,
         long scheduleDelayNanos,
         int maxExportBatchSize,
         long exporterTimeoutNanos,
-        BlockingQueue<ReadableSpan> queue) {
+        ConcurrentLinkedQueue<ReadableSpan> queue,
```

Comment:
I'm trying to get up to speed on thread - did you compare with `ArrayBlockingQueue`? I would be surprised if the linked list, with the allocations it needs to do would perform better and anyways we want to avoid the complicated counter increment. There is almost no work in the blocking section so even with contention it should not really put threads to sleep I guess.

Reply:
I'm particularly worried about the allocations since more garbage means unrelated app code ends up being affected by increased GC

Reply:
could you pease elaborate on what you meant by compare with `ArrayBlockingQueue`? My benchmarks took `ArrayBlockingQueue` the current implementation as the baseline and this approach showed clear improvement on throughput and CPU. Do you mean using `ArrayBlockingQueue` with batched signal? If so, yes I did ran the benchmark but it is much less throughput than using concurrent queue though there are CPU overhead improvements.

Regarding the GC, yes there would be some increased allocations. So ideally we should be pulling in MPSCQueue.

Reply:
> Do you mean using ArrayBlockingQueue with batched signal? I

Yeah I meant this - I don't see it in the graphs, it would be nice to see the numbers to know what the real effect is. If it means simplifying the code (meaning not keeping track of the span count, for example) with a small difference, then it's still worth it especially given we may try to replace with MPSCQueue.

Reply:
Here are the benchmark results. Throughput suffer with BlockingQueue, but exporter CPU overhead is less with batched signal. ConcurrentQueue is better overall though.

![image](https://user-images.githubusercontent.com/62265954/111217253-9ff0a600-8592-11eb-86b0-be7d962de918.png)

![image](https://user-images.githubusercontent.com/62265954/111217270-a3842d00-8592-11eb-8735-02d68f984f2f.png)

I don't see any significant gc.alloc.rate increase. The increase is well within the error boundary.

# With blocking queue batch signal
![image](https://user-images.githubusercontent.com/62265954/111222419-0a0c4980-8599-11eb-87b3-8e13390202c2.png)

# With concurrent queue batch signal
![image](https://user-images.githubusercontent.com/62265954/111222452-17c1cf00-8599-11eb-9a94-b27a91590c1b.png)